rp2040-hal/src/gpio.rs

```diff
@@ -1,132 +1,337 @@
                 }
             }
-        }
 
-        pub struct Parts {
-            $(
-                pub $pxi: $PXi<Input<Floating>>,
-            )+
-        }
+            #[doc = "Struct containing HAL objects for all the "]
+            #[doc = $gpioxs]
+            #[doc = " pins"]
+            pub struct Parts {
+                _pads: pac::$PADSX,
+                _sio: pac::SIO,
+                $(
+                    #[doc = "GPIO pin "]
+                    #[doc = $is]
+                    pub $pxi: $PXi<Unknown>,
+                )+
+            }
+
+            // Puts pad in same state as post-reset, according to datasheet
+            fn setup_pad_io(pads: &pac::$padsx::RegisterBlock, index: usize) {
+                pads.gpio[index].modify(|_, w| w.ie().set_bit()
+                                                .od().clear_bit()
+                                                .pue().clear_bit()
+                                                .pde().set_bit()
+                                                .schmitt().set_bit()
+                                                .drive()._4m_a()
+                                                .slewfast().clear_bit());
+            }
 
-        $(
-            pub struct $PXi<MODE> {
-                _mode: PhantomData<MODE>,
+            fn set_gpio_function(gpios: &pac::$gpiox::RegisterBlock, index: usize, function: u8) {
+                gpios.gpio[index]
+                    .gpio_ctrl
+                    .write_with_zero(|x| unsafe { x.funcsel().bits(function) });
             }
 
-            impl<MODE> $PXi<MODE> {
-                pub fn into_output(
-                    self,
-                    pads: &mut pac::$PADSX,
-                    sio: &mut pac::SIO,
-                ) -> $PXi<Output> {
-                    pads.gpio[$i].modify(|_, w| w.ie().set_bit().od().clear_bit());
-                    unsafe {
-                        (*pac::$GPIOX::ptr()).gpio[$i]
-                            .gpio_ctrl
-                            .write_with_zero(|x| x.funcsel().bits(FUNCTION_SIO));
+            type PacDriveStrength = pac::$padsx::gpio::DRIVE_A;
+
+            $(
+                #[doc = "HAL object for GPIO pin "]
+                #[doc = $is]
+                pub struct $PXi<MODE> {
+                    _mode: PhantomData<MODE>,
+                }
+
+                // Safety: We own our $i slice of padsx, gpiox, and sio because the
+                // construction of Parts assumes ownership of all 3 and will not release
+                // them. Thus several of the methods below will reconstruct these objects
+                // as-needed
+
+                impl<MODE> $PXi<MODE> {
+                    #[doc = "Configure this pin as an output"]
+                    pub fn into_output(
+                        self,
+                    ) -> $PXi<Output> {
+                        unsafe {
+                            setup_pad_io(&*pac::$PADSX::ptr(), $i);
```

Comment:
I would move the dereferences into their own helper methods. That would move all the `unsafe` to one area. In turn all the functions using them would become easier to read.

Reply:
So I'm relatively new to `unsafe`. To me the reason I didn't do that was because I basically didn't want the de-reference to escape the `unsafe` block, since it would be unsafe if we were to use it outside of this context.

Reply:
Oh I see, yes we need to make sure that the reference can't escape. We can use lifetimes to achieve that.
```rust
fn pads<'a>(&'a self) -> &'a pac::pads_bank0::RegisterBlock {
    unsafe {&*pac::$PADSX::ptr()}
}
```

Which ensures that the reference to the register block can at most live as long as the `Gpio$i`.

To make it simpler we can use [lifetime elision](https://doc.rust-lang.org/nomicon/lifetime-elision.html) which happens whenever you take one reference as parameter and return one reference from the function:
```rust
fn pads(&self) -> &pac::pads_bank0::RegisterBlock {
    unsafe {&*pac::$PADSX::ptr()}
}
```

Reply:
Strangely the registers themselves don't seem to require `mut` to write to them. I wonder if that means we should follow the same pattern.

Reply:
Oh yes, that is weird. I would tend to take `&mut` but I'm not sure now. The `OutputPin` trait for example takes `&mut self`.